### ISSUE TITLE
__mul__ method between nn.Module classes raises error

### ISSUE THREAD

**Issue by thomaspinder** [google/flax#636]
I'd like to be able to implement a `__mul__(self, other)` method for classes inheriting the `nn.Module` class. I'm trying to implement a set of kernel functions, and we know that the product of two kernels gives rise to a new kernel, so having the ability to define a new kernel by the product of two existing kernels would be useful. 

A minimal working example that demonstrates the problem is as follows:

```python
class SimpleMLP(nn.Module):
    features: Sequence[int]
    
    @nn.compact
    def __call__(self, inputs):
        x = inputs
        for i, feat in enumerate(self.features):
            x = nn.Dense(feat, name=f'layers_{i}')(x)
            if i != len(self.features) - 1:
                x = nn.relu(x)
        return x

    def __mul__(self, other):
        return self.features + other.features

m1 = SimpleMLP(features=5)
m2 = SimpleMLP(features=10)
m1*m2
```

Clearly the `__mul__(self, other)` method here is only for demonstration purposes, but I would expect `15` to be outputted. Instead, the following error message is given:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-69-ddca13716f29> in <module>
     18 m1 = SimpleMLP(features=5)
     19 m2 = SimpleMLP(features=10)
---> 20 m1*m2

~/Documents/GPBlocks/venv/lib/python3.8/site-packages/flax/linen/module.py in wrapped_module_method(self, *args, **kwargs)
    146 
    147     if self.scope is None:
--> 148       raise ValueError("Can't call methods on orphaned modules")
    149 
    150     if is_compact_method:

ValueError: Can't call methods on orphaned modules
```

Is there a correct way to implement the `__mul__` method for an `nn.Module` class?

**Comment by avital** [comment]
Hi @thomaspinder, good to see you here!

The "Can't call methods on orphaned modules" was supposed to be a helpful error to guard against user error. But this is an interested example of how it may be overly restricting. We could consider alternatively disallowing anything that calls into `self.variables`, `self.variables()` or `self.param()`, which would make your use case work fine.

In the meanwhile, a workaround would be to define `mul` as a top-level function (not a method on your Module instance). I know this isn't ideal but just thought I'd bring it up. I think we should fix this.

**Comment by avital** [comment]
Hi Thomas, here is a PR that (I believe) resolves your issue: https://github.com/google/flax/pull/653

Could you try patching it in and installing a local version of Flax to make sure it indeed covers your use case? Hopefully we can merge this in soon and release a new version on PyPI, but it'd be good to get your feedback first.

**Comment by thomaspinder** [comment]
Yes, this appears to work perfectly for my use case i.e. calling a `__mul__` method. I also tried for `__add__`, and that also appears to be working. Thanks for creating a PR so quickly for this Avi!

**Comment by BertrandRdp** [comment]
Fixed by #653
